File: shared/src/main/java/ru/privatenull/pnauth/kernel/event/AbstractDecisionEvent.java
package ru.privatenull.pnauth.kernel.event;
import java.time.Instant;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
public abstract class AbstractDecisionEvent implements DecisionEvent {
    private final CopyOnWriteArrayList<EventDecision> history = new CopyOnWriteArrayList<>();
    private volatile EventDecision effective = new EventDecision(false, "kernel:default", EventPriority.LOWEST, "", Instant.now());
    @Override public boolean cancelled() { return effective.cancelled(); }
    @Override public void setCancelled(boolean cancelled) { setDecision(cancelled, ""); }
    @Override public synchronized void setDecision(boolean cancelled, String reason) {
        ListenerOptions actor = EventDispatchScope.current();
        if (actor == null) actor = ListenerOptions.normal("external:direct");
        if (actor.mode() == ListenerMode.MONITOR) throw new IllegalStateException("MONITOR listeners are read-only");
        EventDecision proposed = new EventDecision(cancelled, actor.ownerId(), actor.priority(),
                reason == null ? "" : reason, Instant.now());
        history.add(proposed);
        if (actor.priority().value() >= effective.priority().value()) effective = proposed;
    }
    @Override public EventDecision effectiveDecision() { return effective; }
    @Override public List<EventDecision> decisionHistory() { return List.copyOf(history); }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/kernel/event/AbstractDecisionEvent.kt
package ru.privatenull.pnauth.kernel.event

import java.time.Instant
import java.util.concurrent.CopyOnWriteArrayList

abstract class AbstractDecisionEvent : DecisionEvent {
    private val history = CopyOnWriteArrayList<EventDecision>()
    @Volatile
    private var effective = EventDecision(false, "kernel:default", EventPriority.LOWEST, "", Instant.now())

    override fun cancelled(): Boolean = effective.cancelled

    override fun setCancelled(cancelled: Boolean) {
        setDecision(cancelled, "")
    }

    @Synchronized
    override fun setDecision(cancelled: Boolean, reason: String?) {
        val actor = EventDispatchScope.current() ?: ListenerOptions.normal("external:direct")
        check(actor.mode != ListenerMode.MONITOR) { "MONITOR listeners are read-only" }
        val proposed = EventDecision(
            cancelled, actor.ownerId, actor.priority,
            reason ?: "", Instant.now()
        )
        history.add(proposed)
        if (actor.priority.value >= effective.priority.value) {
            effective = proposed
        }
    }

    override fun effectiveDecision(): EventDecision = effective
    override fun decisionHistory(): List<EventDecision> = history.toList()
}
